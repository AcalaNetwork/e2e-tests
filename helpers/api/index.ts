import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
