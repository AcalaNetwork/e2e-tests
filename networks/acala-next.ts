import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
