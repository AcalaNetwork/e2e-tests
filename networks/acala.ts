import { Config } from './types'

type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  dot: { Token: 'DOT' },
  wbtc: { ForeignAsset: 5 },
}

export const karura = {
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
}
