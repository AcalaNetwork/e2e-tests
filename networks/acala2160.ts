import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
