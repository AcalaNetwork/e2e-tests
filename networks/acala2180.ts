import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
