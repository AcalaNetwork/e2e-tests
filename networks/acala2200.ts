import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
