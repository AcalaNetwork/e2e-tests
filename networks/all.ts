import { Config } from './types'

import acala2160Config from './acala2160'
import acalaConfig from './acala'
import hydraDXConfig from './hydraDX'
import moonbeamConfig from './moonbeam'
import polkadot9381Config from './polkadot9381'
import polkadotConfig from './polkadot'
import statemintConfig from './statemint'

const all = {
  polkadot: polkadotConfig,
  polkadot9381: polkadot9381Config,
  statemint: statemintConfig,
  acala: acalaConfig,
  acala2160: acala2160Config,
  moonbeam: moonbeamConfig,
  hydraDX: hydraDXConfig,
} satisfies Record<string, Config>

export default all
