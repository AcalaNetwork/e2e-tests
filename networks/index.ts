import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
