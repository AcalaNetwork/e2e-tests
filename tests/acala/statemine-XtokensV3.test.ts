import { afterAll, beforeEach, describe, it } from 'vitest'
import { connectParachains } from '@acala-network/chopsticks'

import { check, checkEvents, checkHrmp, checkSystemEvents, sendTransaction, testingPairs } from '../helper'
import { xTokensTransferMulticurrenciesV3 } from '../api/extrinsics'
import networks from '../networks'

describe('Karura <-> Statemine', async () => {
  const { alice, bob } = testingPairs()

  const statemine = await networks.statemine()
  const karura = await networks.karura({
    wasmOverride: './wasm/karura-2160-dev.wasm',
  })
  await connectParachains([statemine.chain, karura.chain])

  await karura.dev.newBlock()

  beforeEach(async () => {
    await statemine.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
      Assets: {
        Account: [[[1984, alice.address], { balance: 1000e6 }]],
      },
    })
    await karura.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: 'KSM' }], { free: 10 * 1e12 }],
          [[alice.address, { ForeignAsset: '7' }], { free: 10 * 1e12 }],
        ],
      },
    })
  })

  afterAll(async () => {
    await statemine.teardown()
    await karura.teardown()
  })

  it('Statemine transfer assets to Karura', async () => {
    await statemine.api.tx.polkadotXcm
      .limitedReserveTransferAssets(
        {
          V0: {
            X2: [{ Parent: null }, { Parachain: 2000 }],
          },
        },
        {
          V0: {
            X1: {
              AccountId32: {
                network: 'Any',
                id: alice.addressRaw,
              },
            },
          },
        },
        {
          V0: [
            {
              ConcreteFungible: {
                amount: 10e6,
                id: {
                  X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
                },
              },
            },
          ],
        },
        0,
        { Unlimited: null }
      )
      .signAndSend(alice)

    await statemine.chain.newBlock()

    await checkSystemEvents(statemine, 'polkadotXcm').toMatchSnapshot()
    await checkHrmp(statemine).toMatchSnapshot()
    await check(statemine.api.query.system.account(alice.address)).redact().toMatchSnapshot()

    await karura.chain.newBlock()

    await checkSystemEvents(karura, 'xcmpQueue').toMatchSnapshot()

    // ensure Alice got the money
    await check(karura.api.query.tokens.accounts(alice.address, { ForeignAsset: '7' }))
      .redact()
      .toMatchSnapshot()
  })

  it('Karura transfer assets to Statemine', async () => {
    const tx = await sendTransaction(
      xTokensTransferMulticurrenciesV3(karura.api, '7', '9999199', '1000', bob.addressRaw).signAsync(alice)
    )
    await karura.chain.newBlock()

    await check(karura.api.query.tokens.accounts(alice.address, { ForeignAsset: '7' }))
      .redact()
      .toMatchSnapshot()

    await checkEvents(tx, 'xTokens', 'xcmpQueue').toMatchSnapshot()
    await checkHrmp(karura).toMatchSnapshot()

    await statemine.chain.newBlock()

    await checkSystemEvents(statemine, 'xcmpQueue').toMatchSnapshot()
    await check(statemine.api.query.assets.account(1984, bob.address)).redact().toMatchSnapshot()
  })
})
