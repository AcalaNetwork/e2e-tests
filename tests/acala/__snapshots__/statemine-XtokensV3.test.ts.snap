// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Karura <-> Statemine > Karura transfer assets to Statemine > events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 16000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
        {
          "fun": {
            "Fungible": "(rounded 10000000)",
          },
          "id": {
            "Concrete": {
              "interior": {
                "X3": [
                  {
                    "Parachain": 1000,
                  },
                  {
                    "PalletInstance": 50,
                  },
                  {
                    "GeneralIndex": "(rounded 2000)",
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 1000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 16000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`Karura <-> Statemine > Karura transfer assets to Statemine > hrmp 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "withdrawAsset": [
              {
                "fun": {
                  "fungible": 9999199,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              {
                "fun": {
                  "fungible": 300000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "here": null,
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 300000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "here": null,
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 2,
            },
          },
        ],
      },
    ],
    "recipient": 1000,
  },
]
`;

exports[`Karura <-> Statemine > Karura transfer assets to Statemine > system events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": "(rounded 1300000000)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Karura <-> Statemine > Karura transfer assets to Statemine 1`] = `
{
  "free": "(rounded 10000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Karura <-> Statemine > Karura transfer assets to Statemine 2`] = `
{
  "balance": "(rounded 10000000)",
  "extra": null,
  "isFrozen": false,
  "reason": {
    "sufficient": null,
  },
}
`;

exports[`Karura <-> Statemine > Statemine transfer assets to Karura > hrmp 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x3": [
                        {
                          "parachain": 1000,
                        },
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x3": [
                        {
                          "parachain": 1000,
                        },
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 2000,
  },
]
`;

exports[`Karura <-> Statemine > Statemine transfer assets to Karura > system events 1`] = `
[
  {
    "data": [
      {
        "interior": {
          "X1": {
            "Parachain": 2000,
          },
        },
        "parents": 1,
      },
      3,
    ],
    "method": "SupportedVersionChanged",
    "section": "polkadotXcm",
  },
  {
    "data": [
      {
        "Complete": "(rounded 650000000)",
      },
    ],
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`Karura <-> Statemine > Statemine transfer assets to Karura > system events 2`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "messageHash": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Karura <-> Statemine > Statemine transfer assets to Karura 1`] = `
{
  "consumers": 0,
  "data": {
    "feeFrozen": 0,
    "free": "(rounded 10000000000000)",
    "miscFrozen": 0,
    "reserved": 0,
  },
  "nonce": 1,
  "providers": 0,
  "sufficients": 0,
}
`;

exports[`Karura <-> Statemine > Statemine transfer assets to Karura 2`] = `
{
  "free": "(rounded 10000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;
